### PR TITLE
Do not leak cursors

### PR DESCRIPTION
**Pull request by davidegrohmann**
This PR replaces #9184

### CODE REVIEW

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/AggregationAcceptanceTest.scala [hunk 1]
```diff
@@ -23,12 +23,6 @@ import org.neo4j.cypher.{ExecutionEngineFunSuite, NewPlannerTestSupport}
 
 class AggregationAcceptanceTest extends ExecutionEngineFunSuite with NewPlannerTestSupport {
 
-  test("apa") {
```
Comment (systay) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/AggregationAcceptanceTest.scala:26]: Oops :)

**Comment by systay** [comment]
This change feels much cleaner and easier to understand than the ThreadLocal variation in #9184. 👍 

There are no unit tests for the new ResourceManager, but other than that, LGTM

**Comment by davidegrohmann** [comment]
@systay I have added the tests now

File: community/cypher/cypher-compiler-3.2/src/test/scala/org/neo4j/cypher/internal/compiler/v3_2/spi/QueryContextAdaptation.scala [hunk 2]
```diff
@@ -32,6 +32,8 @@ trait QueryContextAdaptation {
 
   override type EntityAccessor = this.type
 
+  override def resources: CloseableResource = ???
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.2/src/test/scala/org/neo4j/cypher/internal/compiler/v3_2/spi/QueryContextAdaptation.scala:35]: is this safe to leave unimplemented? If so maybe a comment to future readers :)

Oh I see we do a lot of that already, never mind then...

Comment (davidegrohmann) [reply community/cypher/cypher-compiler-3.2/src/test/scala/org/neo4j/cypher/internal/compiler/v3_2/spi/QueryContextAdaptation.scala:35]: Basically almost all the methods in that class are unimplemented, moreover that class is used only in tests. I don't think there is any need to add any explanation.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/ResourceManager.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    if (!resources.remove(resource)) {
+      throw new IllegalStateException(s"$resource is not in the resource set $resources")
+    }
+  }
+
+  override def close(success: Boolean): Unit = {
+    val iterator = resources.iterator()
+    while ( iterator.hasNext )
+    {
+      iterator.next().close()
```
Comment (thobe) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/ResourceManager.scala]: Should we be handling potential exceptions from the `close()` method?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/spi/BeansAPIRelationshipIteratorTest.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    val iterator = new BeansAPIRelationshipIterator(relationships, mock[NodeManager], new ResourceManager)
+
+    iterator.hasNext
+    verify(relationships).hasNext
+    verify(relationships, times(0)).close()
+
+    iterator.close()
+    verify(relationships).close()
+    verifyNoMoreInteractions(relationships)
+  }
```
Comment (thobe) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/spi/BeansAPIRelationshipIteratorTest.scala:54]: How about a test case to ensure that `close()` is idempotent?

**Review by thobe — COMMENTED** [review]
Looks like a nice approach to solving the resource leak hazard.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/spi/BeansAPIRelationshipIteratorTest.scala (hunk 4 above)
Comment (davidegrohmann) [reply community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/spi/BeansAPIRelationshipIteratorTest.scala:54]: Good point, I'll add one.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/ResourceManager.scala (hunk 3 above)
Comment (davidegrohmann) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/ResourceManager.scala]: Yep, you are right we should.

File: community/cypher/cypher-compiler-3.2/src/test/scala/org/neo4j/cypher/internal/compiler/v3_2/spi/QueryContextAdaptation.scala (hunk 2 above)
Comment (pontusmelke) [reply community/cypher/cypher-compiler-3.2/src/test/scala/org/neo4j/cypher/internal/compiler/v3_2/spi/QueryContextAdaptation.scala:35]: 👍

**Comment by systay** [comment]
To be merged straight after release of Neo4j 3.2

**Pull request merged**